src/server/main_service.cc

```diff
@@ -1582,14 +1582,28 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
 
   interpreter->SetGlobalArray("KEYS", eval_args.keys);
   interpreter->SetGlobalArray("ARGV", eval_args.args);
-  interpreter->SetRedisFunc([cntx, this](auto args) { CallFromScript(cntx, args); });
 
   Interpreter::RunResult result;
   optional<ShardId> sid = GetRemoteShardToRunAt(*tx);
-  if (sid.has_value()) {
+  if (tx->GetMultiMode() != Transaction::NON_ATOMIC && sid.has_value()) {
     // If script runs on a single shard, we run it remotely to save hops.
-    pp_.at(sid.value())->Await([&]() { result = interpreter->RunFunction(eval_args.sha, &error); });
+    interpreter->SetRedisFunc([cntx, this](Interpreter::CallArgs args) {
+      // Disable squashing, as we're using the squashing mechanism to run remotely.
+      args.async = false;
+      CallFromScript(cntx, args);
+    });
+
+    boost::intrusive_ptr<Transaction> stub_tx = new Transaction{cntx->transaction};
+    cntx->transaction = stub_tx.get();
+    tx->PrepareSquashedMultiHop(registry_.Find("EVAL"), [&](ShardId id) { return id == *sid; });
```

Comment:
can you please explain here or in the commit description what happens there?
we disable squashing before `CallFromScript(cntx, args)` but call  `PrepareSquashedMultiHop` here. 
what does it do?

Reply:
I suggested him to re-use parts the squashing mechanism instead of implementing a new function. `PrepareSquashedMultiHop` just initializes the transaction with a "fake" command to prepare the hop. In theory its not needed because we have only a single shard and its correctly marked as active by `StartMultiLockedAhead` 

What is disabled is "real" squashing that is responsible for executing commands in parallel.

Reply:
Nice. So how come it works well and using the real transaction did not work? where do we switch to the original one to allow a command execution? (Beforehand ScheduleSingleHop calls crashed because cb_ptr was already set with the coordinator callback

Reply:
We tried different things 🙂 

Transaction is not designed to be re-entrant, meaning you can't use "itself" while executing a callback. So a custom function needs to be implemented that is not ScheduleSingleHop... But then you run into trouble again, because you run from the callback and multi command logic still submits callbacks and calls PollExecution, yet PollExecution is also not re-entrant, so you run into the weirdest bugs there

In general I'd suggest not using a complicated approach because its difficult to verify and test. We already have squashing that works fairly reliably (and in a much simpler way by offloading all the difficulties to a stub transaction)

Reply:
The issue is that `ScheduleSingleHop` (and many other functions of `Transaction`) are not re-entrant, and so attempts to call them from within a callback calling them requires subtle changes and hacks.
But Vlad correctly commented that the squashing mechanism already handles a similar case, and if a transaction is set to have a `SQUASHED_STUB` role then it executes a different path, similar to what we're trying to achieve:
https://github.com/dragonflydb/dragonfly/blob/02fff36e3e292baf0dc1b4d10680d2830efe2e4f/src/server/transaction.cc#L654

Reply:
Ok, please add statistics about how many times we offload the coordinator vs we do it in its original thread.

Reply:
Just to make sure we do not forget it. For OOO transactions - we can run them immediately and this is something we must check with this code.

Reply:
They will run immediately, but we have +1 hop for scheduling and +1 hop for unlocking

Reply:
we can simulate OOO conditions easily using memtier - like I described to you offline using `redis.call('set', KEYS[1], 'foo'")` script with large key range. In fact, I would like us to repeat the same experiment and also see we get a similar performance as I did.

Reply:
Re/ simulating OOO:

Without this change (i.e. with release 1.9):

```
ALL STATS
===================================================================================================
Type          Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
---------------------------------------------------------------------------------------------------
Evalshas    451195.23         0.55021         0.47900         1.50300         2.30300     46216.21 
Totals      451195.23         0.55021         0.47900         1.50300         2.30300     46216.21 
```

With this change:

```
ALL STATS
===================================================================================================
Type          Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
---------------------------------------------------------------------------------------------------
Evalshas    620400.31         0.40318         0.36700         0.85500         1.77500     63548.00
Totals      620400.31         0.40318         0.36700         0.85500         1.77500     63548.00
```